src/lib/LLVM/HEAD/JIT.hs

```diff
@@ -61,55 +62,68 @@ type CompilationPipeline = LLVM.Module -> IO ()
-              LLVM.AST.GlobalVariable{
-                name="llvm.global_dtors",
-                initializer=Just (C.Array _ elems)} -> elems
-            _ -> []
-      -- Sort in the order of decreasing priority!
-      fmap snd $ sortBy (flip compare) $ flip fmap dtorStructs $
-#if MIN_VERSION_llvm_hs(15,0,0)
-        \(C.Struct _ _ [C.Int _ n, C.GlobalReference (LLVM.AST.Name dname), _]) ->
-#else
-        \(C.Struct _ _ [C.Int _ n, C.GlobalReference _ (LLVM.AST.Name dname), _]) ->
-#endif
-          (n, C8BS.unpack $ SBS.fromShort dname)
 {-# SCC compileModule #-}
 
 foreign import ccall "dynamic"
   callDtor :: FunPtr (IO ()) -> IO ()
 
+loadNativeModule :: JIT -> ExplicitLinkMap -> [String] -> ObjectFileContents -> IO NativeModule
+loadNativeModule moduleJIT linkingMap dtors objFileContents = do
+  moduleDylib <- newDylib moduleJIT
+  let cl = compileLayer moduleJIT
+  addExplicitLinkMap cl moduleDylib linkingMap
+  OrcJIT.addDynamicLibrarySearchGeneratorForCurrentProcess cl moduleDylib
+  moduleDtors <- mapM (lookupDtor moduleJIT moduleDylib) dtors
+  loadObjectFile moduleJIT moduleDylib objFileContents
+  return NativeModule{..}
+
+-- Unfortunately the JIT layers we use here don't handle the destructors properly,
+-- so we have to call them ourselves.
+lookupDtor :: JIT -> OrcJIT.JITDylib -> String -> IO (FunPtr (IO ()))
+lookupDtor JIT{..} moduleDylib dtorName = do
+  Right (OrcJIT.JITSymbol dtorAddr _) <-
+    OrcJIT.lookupSymbol session compileLayer moduleDylib $ fromString dtorName
+  return $ castPtrToFunPtr $ wordPtrToPtr dtorAddr
+
+addExplicitLinkMap :: OrcJIT.IRCompileLayer -> OrcJIT.JITDylib -> ExplicitLinkMap -> IO ()
+addExplicitLinkMap l dylib linkMap = do
+  let (linkedNames, linkedPtrs) = unzip $ M.toList linkMap
+  let flags = OrcJIT.defaultJITSymbolFlags { OrcJIT.jitSymbolAbsolute = True }
+  let ptrSymbols = [OrcJIT.JITSymbol (ptrToWordPtr ptr) flags | ptr <- linkedPtrs]
+  withMangledSymbols l (map fromString linkedNames) \linkedNames' -> do
+    OrcJIT.defineAbsoluteSymbols dylib $ zip linkedNames' ptrSymbols
+
+withMangledSymbols :: OrcJIT.IRLayer l => l -> [SBS.ShortByteString] -> ([OrcJIT.MangledSymbol] -> IO a) -> IO a
+withMangledSymbols _ [] cont = cont []
+withMangledSymbols l (s:ss) cont =
+  OrcJIT.withMangledSymbol l s \ms ->
+    withMangledSymbols l ss \mss -> cont (ms:mss)
+
 -- TODO: This might not release everything if it fails halfway
 unloadNativeModule :: NativeModule -> IO ()
 unloadNativeModule NativeModule{..} = do
   -- TODO: Clear the dylib
   forM_ moduleDtors callDtor
 {-# SCC unloadNativeModule #-}
 
-withNativeModule :: JIT -> [ObjectFileContents] -> LLVM.AST.Module -> CompilationPipeline -> (NativeModule -> IO a) -> IO a
-withNativeModule jit objs m p = bracket (compileModule jit objs m p) unloadNativeModule
+withNativeModule
+  :: JIT -> ExplicitLinkMap -> [String]
+  -> LLVM.AST.Module -> CompilationPipeline -> (NativeModule -> IO a) -> IO a
+withNativeModule jit linkMap dtors m p =
```

Comment:
This shouldn't be taking the extra destructors argument since we can recover those form the LLVM.AST.Module. This is duplicating the data and might end up being error prone in case they disagree